crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -7,18 +7,18 @@ use thiserror::Error;
 
 /// A small wrapper for [`BoxedSystem`] that also keeps track whether or not the system has been initialized.
 #[derive(Component)]
-struct RegisteredSystem {
+struct RegisteredSystem<I, O> {
```

Comment:
This may significantly increase the number of registered components within a World. Every new input-output signature registered in this way will result in a new Component type and ComponentId being allocated.

Reply:
Agreed, I would really like to type-erase this.

Reply:
I don't see how inputs and outputs can be done without type information. 

What is the harm in having a lot of registered components?

Reply:
Hmm okay, I see your point.

Reply:
Yeah, at least for me, I'm fine if this is merged with the existing architecture. Like you said, it's not public, this is incredibly useful functionality, and I don't think it's likely to be a serious performance footgun.

Reply:
> What is the harm in having a lot of registered components?

I haven't measured it, but bloating the components metadata stores may have a negative impact on TypeId -> ComponentId lookups, which most prominently show up in the typed `World::get` and `Entity{Ref/Mut}::get` APIs, which may cause Command slowdowns. This previously wasn't too big of a concern given that the component set typically doesn't grow organically over time after initialization, or if it does, it's not reliant on the generic APIs (i.e. the `*_by_id` untyped APIs), but these changes might allow that to happen over the lifetime of a World.

I'll be a bit more clear and say that I don't think this is blocking. If we're treating `SystemId` as a temporary holdover, then this shouldn't be a huge issue, and I eventually want to find alternatives to that `HashMap<TypeId, ComponentId>` as well.

Reply:
FYI another way this could be done would be adding a supertrait to `System` that's free of the `In` and `Out` associated types, and whose `run_erased` function takes two `&mut dyn Any`:

- the first one is supposed to point to a `Option<I>` and will be `.take()`n by the `run_erased` function to get its input;
- the second one is supposed to point to a `Option<O>` and will be filled by the `run_erased` function with its output.

Then `run_erased` internally knows what the `I` and `O` types were supposed to be and can do the needed downcasts. This way any additional boxing is avoided, and only a dynamic check that the input and output types match up is performed when downcasting the `&mut dyn Any`s.